src/foundations/components/marquee/marquee.tsx

```diff
@@ -0,0 +1,170 @@
+"use client";
+
+import {
+  Children,
+  cloneElement,
+  ComponentPropsWithoutRef,
+  ReactElement,
+  useEffect,
+  useMemo,
+  useRef,
+  useState,
+} from "react";
+
+import { useIntersectionObserver } from "@/foundations/hooks/use-intersection-observer/use-intersection-observer";
+import { useTicker } from "@/foundations/hooks/use-ticker/use-ticker";
+import { getCssValueLength } from "@/foundations/utils/dom/get-css-value-length";
+import { cn } from "@/lib/utils";
+
+type DurationProp = number | ((contentLength: number) => number);
+
+interface MarqueeProps extends ComponentPropsWithoutRef<"div"> {
+  direction?: "left" | "right" | "up" | "down";
+  paused?: boolean;
+  duration?: DurationProp;
+}
+
+export const Marquee = ({
+  direction: propDirection = "left",
+  children,
+  paused = false,
+  duration = (l) => l * 0.05,
+  style,
+  className,
+  ...props
+}: MarqueeProps) => {
+  const rootRef = useRef<HTMLDivElement>(null);
+
+  const [numClones, setNumClones] = useState<number>(1);
+  const { isIntersecting } = useIntersectionObserver(rootRef);
+
+  const progress = useRef(0);
+  const contentLength = useRef(0);
+  const deferredResizeHandler = useRef<() => void | null>(null);
+
+  const [axis, direction] = useMemo(() => {
+    return [
+      propDirection === "up" || propDirection === "down" ? "y" : "x",
+      propDirection === "up" || propDirection === "left" ? "normal" : "reverse",
+    ];
+  }, [propDirection]);
+
+  const ticker = useTicker((timestamp, delta) => {
+    if (deferredResizeHandler.current) {
+      deferredResizeHandler.current();
+      deferredResizeHandler.current = null;
+    }
+
+    const root = rootRef.current;
+    if (!root) return;
+
+    const durationValue =
+      typeof duration === "number" ? duration : duration(contentLength.current);
+
+    progress.current =
+      (progress.current + delta / (durationValue * 1000)) % 1 || 0;
+
+    root.style.setProperty("--p", progress.current.toString());
```

Comment:
nitpick, but I would make the css custom properties a bit more descriptive as I'm having a bit of a bad time trying to understand what they stand for with just one letter